clevertap-core/src/main/java/com/clevertap/android/sdk/inapp/evaluation/TriggersMatcher.kt

```diff
@@ -0,0 +1,404 @@
+        }
+    }
+
+    /**
+     * Internal function to evaluate a haversine distance condition.
+     *
+     * This function evaluates if the haversine distance between two locations is within a specified radius.
+     * The haversine formula is used to compute the distance between two locations.
+     *
+     * @param radius The radius to check against, in kilometers.
+     * @param expected The expected location.
+     * @param actual The actual location.
+     * @return `true` if the haversine distance is within the specified radius, `false` otherwise.
+     */
+    @VisibleForTesting
+    internal fun evaluateDistance(radius: Double, expected: Location, actual: Location): Boolean {
+        val distance = Utils.haversineDistance(expected, actual)
+        return distance <= radius
+    }
+
+    /**
+     * Internal function to check if the expected value equals the actual value.
+     *
+     * This function checks if the expected value is equal to the actual value. It handles various
+     * data types including strings, numbers, and lists.
+     *
+     * @param expected The expected [TriggerValue].
+     * @param actual The actual [TriggerValue].
+     * @return `true` if the expected value equals the actual value, `false` otherwise.
+     */
+    @VisibleForTesting
+    internal fun expectedValueEqualsActual(expected: TriggerValue, actual: TriggerValue): Boolean {
+        return when {
+            expected.isList() && actual.isList() -> {
+                expected.listValueWithCleanedStringIfPresent()!!
+                    .toHashSet() == actual.listValueWithCleanedStringIfPresent()!!.toHashSet()
+            }
+
+            actual.isList() ->
+                checkGivenElementEqualsAnyElementInList(
+                    actual.listValueWithCleanedStringIfPresent()!!,
+                    expected.value
+                )
+
+            expected.isList() -> checkGivenElementEqualsAnyElementInList(
+                expected.listValueWithCleanedStringIfPresent()!!,
+                actual.value
+            )
+
+            expected.numberValue() != null -> {
+                val actualNumber =
+                    actual.numberValue()?.toDouble() ?: actual.stringValueCleaned()?.toDoubleOrNull()
+                    ?: return false
+                expected.numberValue()!!.toDouble() == actualNumber
+            }
+
+            actual.numberValue() != null -> {
+                val expectedNumber =
+                    expected.stringValueCleaned()?.toDoubleOrNull() ?: return false
+                actual.numberValue()!!.toDouble() == expectedNumber
+            }
+
+            actual.stringValue() != null -> expected.stringValueCleaned() == actual.stringValueCleaned()
+            else -> false
+        }
+    }
+
+    @VisibleForTesting
+    internal fun expectedValueLessThanGreaterThanActual(
+        expected: TriggerValue,
+        actual: TriggerValue,
+        isLessThan: Boolean
+    ): Boolean {
```

Comment:
investigate if there is a better way of handling nulls, only one place should handle null values and have practically all defaults set

Reply:
tech debt